### DISCUSSION TITLE
Make your requests to run tests on a *real* S3 Trio64V2/DX here!

### DISCUSSION THREAD

**Discussion by iamgreaser** [86Box/86Box#3898] [General]
I finally got my HP Brio 8314 running with a VGA cable plugged into a screen. It's running an older video BIOS than the one we have in 86Box already (v2.04.07 vs v2.04.11), and once I have a good way to transfer stuff to and from it, I'll be uploading that and then getting some test programs together for reproducing various VGA behaviours and whatnot. (It's got a D-Link NE2000 clone but it's ISA-PnP which truly is the spawn of Satan, and will require several reboots until I finally correctly guess what IRQ the BIOS has chosen today.)

Behaviours MUST be properly narrowed down to a small test case I can run in `debug`, rather than "this specific flight simulator that comes on 10 CDs that you install to your hard drive does a weird thing 20 minutes in" - I don't want to fill my hard drive with crap thanks. Having said that, Commander Keen 4 works perfectly without needing the SVGA compatibility fix, so at least in EGA modes the 256 KB wraparound should work properly.

The video chip is an S3 Trio64V2/DX and there is 2 MB of VRAM onboard.

BEHAVIOURS WHICH SHOULD BE CONSISTENT WITH A REAL VGA:

- 2x2bpp semichunky mode works perfectly, as with everything else which isn't an emulator. (gc.0x05.b6=0,b5=1)
- The color/mono bit in the attribute controller (attr.0x10.b1) selects which kind of graphics blink mode to use - color mode means plane 3 has to be set to 1 to make it blink, and 0 to force the plane 3 output to 1, while mono mode means plane 3 is always hooked up to blink. The output is still in colour.
- It handles 8bpp graphics blink modes. That is, the low 4 bits and the high 4 bits act as independent sets of planes 0 through 3 for the purposes of blink.
- The sequencer "load every 1/2/4 character clocks" setting is honoured (seq.0x01.b2/b4).
- In a 9-dot text mode, if one loads every 2 or 4 character clocks (seq.0x01.b2/b4), and block characters are enabled (attr.0x10.b2=1), then the rightmost column side is repeated for the extra character clocks on the relevant characters, and background colour for the other characters.

BEHAVIOURS WHICH ARE PROBABLY WRONG:

- The 8bpp enable bit in the attribute controller (attr.0x10.b6=1) makes it ignore every second dot clock or something, repeating the previous pixel.
- The 8bpp-compatible-shifter bit in the graphics controller (gc.0x05.b6=1) enables 8bpp mode, although you get some weird "we shift in 4 bits each dot clock and you get to see some weird intermediate colours happening" behaviour.

WEIRD BEHAVIOURS:

- I'll need to look into how the EGA colour palette interacts with the VGA colour palette, but you do get some funky stuff happening there when playing around with hybrid-4bpp-8bpp stuff.
- When loading every 2 or 4 character clocks, 1s get shifted into the shift registers, so you effectively end up with plane 3 being filled with 0xFF.
- Based on experiments with the "load every 1/2/4 character clocks" setting, the character skew appears to read the first "character" multiple times until it's time to actually start reading from the correct location. What actually happens with a 1-char skew with "load every 4 clocks" in mode 0x12 (640x480x4bpp) is that you see the first column after one full plane shift, then the next 2 plane shifts, and then you finally see the first column *again* with the correct colours... and then you see the next 3 plane shifts of the first column, then you get to the second column and all its plane shifts, then the third...
- If one enables the CGA compatibility shifter setting (as used for 2x2 semichunky - gc.0x05.b6=0,b5=1) in a text mode, plane 3 data is also incorporated. This is inconsistent with an EGA, but maybe a VGA would work this way?

SVGA SPECIFICS:

- Linear mode is activated via crtc.0x58.b4=1. This can also be applied in text mode and the results are... well, linear. At least the text looks sane.
- High-res 8bpp mode is activated via crtc.0x3A.b4=1. This forces the data to be interpreted as 8bpp graphics. When one applies this in mode 13h, the screen looks somewhat doubled and also seems to be shifted slightly to the left.
- (EDIT: There's probably some other things going on here - this was what I got when setting an SVGA mode.) The colour mode field in crtc.0x67.b4...7, when set to something other than 0, visually overrides the linear mode flag (crtc.0x58.b4) treating it as set (possibly also affecting memory access?), and also *mostly* overrides the high-res 8bpp mode flag (crtc.0x3A.b4) treating it as also set. The catch is that clearing the latter seems to blank the left few pixels or something like that.
- If one wants to use doubled-8bpp, 15bpp, or 16bpp as per crtc.0x67.b4...7, one also needs to set seq.0x15.b6=1 to halve the clock speed, otherwise good luck getting your monitor to sync to a 140 Hz signal. (...this message was posted on a computer with a 144 Hz-capable monitor.) The results in mode 13h, after clearing the attr 8bpp mode flag (attr.0x10.b6=0), are surprisingly readable... and doubled horizontally, and the character skew is incorrect, and for 16bpp the text is green (and in theory with a hint of blue). This will require further investigation, but it strongly hints that a lot of the VGA compat stuff is still intact even if you tell it to use 15/16/24/32bpp output.

**Comment by iamgreaser** [comment]
At least after unlocking a bunch of registers (CRTC 0x4838, 0xA539, reg 0x40 is ORed with 0x01, SEQ 0x0608), the following is true for register reads:

Sequencer:
- Only the low 6 bits of the address is used.
- Registers 0x10 through 0x1F are mirrored 2 more times through 0x20 through 0x3F.
- If a register does not exist, the resulting value is the low 6 bits of the address, with the high 2 bits forced to 0s. This applies even for the mirrored registers.

Graphics Controller:
- Only the low 4 bits of the address are used.
- If a register does not exist, the resulting value is the low 4 bits of the address, with the high 4 bits ~~forced to 1s.~~ The high 4 bits come from the high 4 bits of the bit mask register (0x08).
- This includes the 0xF8 through 0xFB range, which according to some spec is supposed to return the 4 bytes of the latch... according to some comment in `vid_svga.c`. Not sure which spec that is, but this chip definitely doesn't follow it!

CRTC:
- All 8 bits of the address are used.
- If a register does not exist, the resulting value is the last value read from or written to a *valid* register.

I haven't tested the attribute controller yet as, well, if you've ever tried to program for the attribute controller... OK, I will say one thing about it:
- Reading 0x3DA resets the address/data flip-flop. Writing to 0x3C0 sets the flip-flop. Reading from 0x3C1 does not reset the flip-flop, so it's somewhat convenient for read-modify-write stuff. (Seriously IBM, why did you decide to be obnoxiously and pointlessly different with the attribute controller?) This is probably standard VGA behaviour. Not sure if writing sets or toggles, I will have to check this at some stage.

Here's a map of registers which exist in the CRTC space. MSbit is first, LSbit is last - so the first byte refers to CR00 in the MSbit, and CR07 in the LSbit.
```
FF FF FF 80 70 07 FF 38
D5 77 FF EE 95 FF 22 00
FA 80 F0 00 00 00 00 00
00 00 00 00 00 00 00 00
```

And no, the 0x80-0x84 space (I could have sworn what was in 0x85 was valid though...?!) is NOT the same as the S3 Trio64UV+ - it actually contains a 0x10, the PCI vendor (16-bit little-endian), and the PCI product (16-bit little-endian).

**Reply by iamgreaser** [reply]
Attribute Controller reads (NOTE: tested without unlocking anything):

- Only the low 6 bits of the address is used.
- If a register does not exist, the resulting value is the low 6 bits of the address, with the high 2 bits forced to 0s.
- Accesses of 0x20 through 0x2F are treated as nonexistent.
- 0x18 through 0x1F are mirrors of 0x10 through 0x17, and the same logic applies for 0x3x.
- 0x10 through 0x14 exist, 0x15 through 0x17 do not.

**Reply by iamgreaser** [reply]
Reading index registers after writing them:

- Sequencer: Binary 00xxxxxx.
- Graphics Controller: ~~Weird. 0x05 reads back as 0x15, otherwise sign-extended 4 bits.~~ Wrong, see later comment for actual behaviour.
- CRT Controller: All 8 bits as-is.
- Attribute Controller: Binary 00xxxxxx.

**Reply by iamgreaser** [reply]
UPDATE: Graphics Controller behaves differently from what I expected. Other things likely to behave different too, but here's what I have so far for just the GC:

- Write index: Only the low 4 bits of the index are used.
- Read index: Low 4 bits return from the index, high 4 bits copied from the actual register. But if this value is out of range, the register that the data is taken from is register 0x08 (bit mask).
- Read data: Read if 0x00 <= (index & 0x0F) <= 0x08, otherwise behave like "read index".

**Comment by eddmanx** [comment]
Is your card VBE 2.0? If so, could you test the VBE display mode of Duke 3D, at 640x480 or higher, and see if there are slowdowns with mirrors?

You can enable the fps counter by typing DNRATE.

**Reply by iamgreaser** [reply]
Yes, it's VBE 2.0. Was going to say no but my machine could do with having Duke 3D on it so might as well test.

To clarify, this is running on a Covington - 266MHz Celeron, probably with a 66 MHz bus.

The first few columns on the left are clipped off and I have actually adjusted my screen so I could feasibly see them if they somehow weren't.

I get about 18 FPS on the bathroom mirror on E1L1. The same happens if I write a TSR to suppress all calls to INT 0x10 AX=0x4F07, but I suspect that's not the workaround one is supposed to use...

**Reply by eddmanx** [reply]
Thanks. The fps does seem to be low for that processor and card.

Could you try it with MSKVBEF7? https://www.vogons.org/viewtopic.php?f=63&t=57420

It massively increases the fps for VBE 2.0+ cards, at least in 86box.

The TSR that you wrote, I suppose it mostly acts like the NOLFB TSR?

**Reply by iamgreaser** [reply]
(Shoutouts to DuckDuckGo for making it a pain to re-find that. I managed to find it by searching for "duke3d nvidia vogons" and it was on page 3. Seriously, search sucks these days.)

I actually went ahead and tried a whole bunch of different things, including the equivalent of MSKVBEF7, and it genuinely seems to have no real impact on performance.

I'm rewriting this to do only what MSKVBEF7 does and also to use the "proper" DOS way of hooking an interrupt rather than hooking the interrupt table directly and it's giving me grief... I'll push through anyway and get you some results.

**Reply by iamgreaser** [reply]
...OK, when uninstalling my TSR I was resetting the INT 10h with an uninitialised far pointer to what would be the normal INT 10h handler within the uninstalling process's segment instead of grabbing it from the running TSR process's chain far pointer, so when it tried to print the "Uninstalled successfully!" message it would hang the system. Oops!

Anyway, I reduced the TSR down to what MSKVBEF7 does (and yes I actually looked at the source code), and... no difference in performance.

**Reply by eddmanx** [reply]
I wonder why it doesn't improve the performance. Why does it work in 86box then?

**Reply by iamgreaser** [reply]
There's a few possibilities, the obvious possibility being "it's not accurate enough", but another one is that I have an older video BIOS.

Here's the video BIOS I'm using, try swapping your one out in 86Box and seeing if the FPS still improves: [s3trio64v2dx-2-04-07.zip](https://github.com/86Box/86Box/files/13799310/s3trio64v2dx-2-04-07.zip)

On the other hand I could try shadowing in the one that's in 86Box - the board has an i440EX + PIIX4E and I have datasheets for those.

**Reply by eddmanx** [reply]
Your bios behaves the same; fps is improved drastically.

Now if the reason for it working is because of inaccuracy, I'd rather see it stay that way then. :grin:

**Reply by iamgreaser** [reply]
A better solution would be to make it accurate and then tell people to emulate something faster :P

Anyway, managed to shadow in and initialise the 2.04.11 BIOS from 86Box and the FPS is mostly unaffected.

That was a somewhat disgusting experience to implement for real - I had to drop in a stub INT 0x10 handler... and calling the video BIOS initialisation function while FreeDOS is running in protected mode with a DPMI server running just feels like so many levels of wrong...

But today has been so many levels of wrong already that I'm starting to wonder if I should pinch myself just to see if I'm still dreaming.

**Reply by iamgreaser** [reply]
...OK, following up on this.

I managed to run a S3 Trio64 video BIOS as a TSR (specifically, 86Box's 86c764x1.bin). It's a bit temporamental as to whether it will successfully set the mode or decide to just hang, but when it does end up doing a 640x480 mode... the performance is awful, and I lose *another* column on the left so I can't see the first digit of the FPS counter at all. Still, it seems to get about 18 FPS at the mirror.

I did confirm that it's a VBE 1.2 video BIOS.

... and for completeness: No, you cannot run an S3 Trio64V2/DX off an IBM VGA BIOS. It does not work.

**Reply by iamgreaser** [reply]
Well, it turns out I screwed up here and accidentally made the TSR hook of a call to INT 0x10 AX=0x4F01 return immediately if it was *successful*. Inverting the check allows me to conclude the following:

**Reducing it down to a single buffer and/or not using the LFB actually makes general performance *worse* on an S3 Trio64V2.**

For completeness, I should have mentioned that I'm using the Steam release of 1.5 Atomic Edition, and all this time I have been using 640x480.

**Reply by eddmanx** [reply]
According to the dev of mskvbef7 reducing to single buffer but keeping LFB should help with slowdowns, well at least not hurt it, and that's on real hardware.

https://www.vogons.org/viewtopic.php?p=1141494#p1141494
https://www.vogons.org/viewtopic.php?p=1221796#p1221796

I'm not an expert on these things and not a programmer. Maybe contact the dev on that thread; see what might be different with your card or system.

**Comment by iamgreaser** [comment]
Memory layout of mode 13h, as captured via SVGA mode 81FFh:

- 4 bytes of a VGA plane pertaining to the mode 13h screen
- 4 bytes of 0x00
- 4 bytes of something unused mixed with font data (set to 0xFF 0xFF fontbyte 0x00)
- 4 bytes of 0x00
- 4 bytes of the previously used text screen (set to char attr fontbyte 0x00)
- 4 bytes of 0x00
- 4 bytes of something unused mixed with font data (set to 0xFF 0xFF fontbyte 0x00)
- 4 bytes of 0x00

The same, according to 86Box at the time of writing (and also DOSBox and DOSBox-X for that matter):

- A packed pixel mode 13h screen

DEBUG program used for testing (0x31 bytes):
```
a100
mov ax, 13
int 10
mov ax,a000
mov es, ax
xor di, di
mov al, 0
stosb
inc al
jnz 10e           ; points to the "stosb" above
mov ax,4f02
mov bx,81ff
int 10
push cs
push es
pop ds
pop es
xor si,si
mov di,1000
mov cx,1000
rep movsb
push cs
pop ds
mov ax,3
int 10
int 3

g=100
d1000
```

**Reply by iamgreaser** [reply]
After setting CR68.7 to 0, which forces the chip to think I've got a 32-bit bus (and also reduces the VRAM to 1 MB), the memory layout starts looking more like this:
- 4 bytes of a VGA plane pertaining to the mode 13h screen
- 4 bytes of something unused mixed with font data (here, set to 0x00 0x00 fontbyte 0x00)
- 4 bytes of the previously used text screen (set to 0xFF 0xFF fontbyte 0x00)
- 4 bytes of something unused mixed with font data (here, set to 0x00 0x00 fontbyte 0x00)

Then I set CR68.7 back to 1 and it seems to have filled the upper 32 bits of each 64-bit word with... some weird internal stuff? Or maybe the data contents of the upper 32-bit pair of RAM chips is decaying to noise?
